### ISSUE TITLE
Possibility of a New Authentication Method?

### ISSUE THREAD

**Issue by rstad944** [vloschiavo/powerwall2#55]
I noticed in the latest Tesla Android app under Settings that there is an option to "Pair Phone to Powerwall" which requires a toggle of the PW switch.

Has anyone been able to exploit this option to pair a local controller (e.g. Raspberry Pi) to the TEG via a LAN connection?

I am especially wondering if the Powerwall mode (backup, self powered) could once again be controlled locally without using the owner-api.teslamotors.com endpoint.  This local control capability had been essentially removed when the PW firmware was updated from 1.49 to 1.50.

**Comment by dan-simpson-zip** [comment]
Also keen to hear more about this!  Just tried implementing a POST to `/api/operation` in tesla_powerwall python library repository, and it wasn't till I ran an integration test that I happened upon an access denied error, e.g.:

```
tesla_powerwall.error.AccessDeniedError: Access denied for resource /api/operation: Unable to POST to resource: User does not have adequate access rights
```

Would be fantastic to modify backup percentage via api 🙏

**Comment by daniel-simpson** [comment]
Also keen to hear more about this!  Just tried implementing a POST to `/api/operation` in tesla_powerwall python library repository, and it wasn't till I ran an integration test that I happened upon an access denied error, e.g.:

```
tesla_powerwall.error.AccessDeniedError: Access denied for resource /api/operation: Unable to POST to resource: User does not have adequate access rights
```

Would be fantastic to modify backup percentage via api 🙏

**Comment by vloschiavo** [comment]
Agree. This does sound very interesting.  I'll leave this issue open until we can find a solution.

**Comment by patniemeyer** [comment]
I was able to make API calls to my gateway by grabbing the bearer token after logging in to the local web app (as "customer").  It only lasts for a few hours but it allows me to make calls to the /vitals endpoint.

**Comment by longzheng** [comment]
> I was able to make API calls to my gateway by grabbing the bearer token after logging in to the local web app (as "customer"). It only lasts for a few hours but it allows me to make calls to the /vitals endpoint.

@patniemeyer Just double checking if you're still able to replicate this? I tested using a bearer token and am still seeing a 404 for `/api/devices/vitals`.

```
GET /api/devices/vitals HTTP/1.1
Authorization: Bearer REDACTED
User-Agent: PostmanRuntime/7.36.1
Accept: */*
Host: 192.168.1.68
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
 
HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Tue, 23 Jan 2024 07:37:15 GMT
Content-Length: 19
 
404 page not found
```

**Comment by Sn0w3y** [comment]
@longzheng as we can see it is REDACTED.

Maybe we can implement a Authentication to the Code?

Greetings

**Comment by Matthew1471** [comment]
I'm working on this.. it involves a lot of protobuffs and generating encryption/decryption keys. The power toggle is to "accept" a public encryption key.

This is in line with the vehicles too, so some more info is [here](https://github.com/teslamotors/vehicle-command/blob/main/pkg/protocol/protocol.md) and [here](https://github.com/teslamotors/vehicle-command/blob/main/README.md)

**Comment by Matthew1471** [comment]
Working examples here on how to use the LAN API. "gateway_*" are Tesla Gateway source examples: https://github.com/Matthew1471/Tesla-API/blob/main/Documentation/Wrappers/Python/README.adoc#examples
